src/utils/generator.js

```diff
@@ -52,12 +55,159 @@ import type {
+  | "ABSTRACT_OBJECT_SET_PARTIAL"
+  | "ABSTRACT_OBJECT_SET_PARTIAL_VALUE"
+  | "ABSTRACT_OBJECT_GET_PARTIAL"
+  | "ABSTRACT_OBJECT_GET_PROTO_OF"
+  | "ABSTRACT_OBJECT_GET"
+  | "DIRECT_CALL_WITH_ARG_LIST"
+  | "CALL_ABSTRACT_FUNC"
+  | "CALL_ABSTRACT_FUNC_THIS"
+  | "CALL_BAILOUT"
+  | "EMIT_CALL"
+  | "EMIT_CALL_AND_CAPTURE_RESULT"
+  | "GET_BINDING"
+  | "LOCAL_ASSIGNMENT"
+  | "LOGICAL_PROPERTY_ASSIGNMENT"
+  | "CONDITIONAL_PROPERTY_ASSIGNMENT"
+  | "PROPERTY_ASSIGNMENT"
+  | "MODULES_REQUIRE"
+  | "RESIDUAL_CALL"
+  | "ASSUME_CALL"
+  | "CANNOT_BECOME_OBJECT"
+  | "UPDATE_INCREMENTOR"
+  | "WIDENED_IDENTIFIER"
+  | "WIDEN_PROPERTY"
+  | "WIDEN_ABSTRACT_PROPERTY"
+  | "WIDEN_PROPERTY_ASSIGNMENT"
+  | "WIDEN_ABSTRACT_PROPERTY_ASSIGNMENT"
+  | "INVARIANT"
+  | "INVARIANT_APPEND"
+  | "DERIVED_ABSTRACT_INVARIANT"
+  | "PROPERTY_INVARIANT"
+  | "FULL_INVARIANT"
+  | "FULL_INVARIANT_FUNCTION"
+  | "FULL_INVARIANT_ABSTRACT"
+  | "UNKNOWN_ARRAY_METHOD_CALL"
+  | "UNKNOWN_ARRAY_METHOD_PROPERTY_CALL"
+  | "UNKNOWN_ARRAY_LENGTH"
+  | "UNKNOWN_ARRAY_GET_PARTIAL"
+  | "BABEL_HELPERS_OBJECT_WITHOUT_PROPERTIES"
+  | "REACT_DEFAULT_PROPS_HELPER"
+  | "REACT_TEMPORAL_FUNC"
+  | "REACT_CREATE_CONTEXT_PROVIDER"
+  | "REACT_SSR_REGEX_CONSTANT"
+  | "REACT_SSR_PREV_TEXT_NODE"
+  | "REACT_SSR_RENDER_VALUE_HELPER"
+  | "REACT_SSR_TEMPLATE_LITERAL"
+  | "REACT_NATIVE_STRING_LITERAL"
+  | "REACT_RELAY_MOCK_CONTAINER"
+  | "FB_MOCKS_BOOTLOADER_LOAD_MODULES"
+  | "FB_MOCKS_MAGIC_GLOBAL_FUNCTION";
+
+export type DerivedExpressionBuildNodeFunction = (
+  Array<BabelNodeExpression>,
+  SerializationContext,
+  Set<AbstractValue | ObjectValue>
+) => BabelNodeExpression;
+
+export type ResidualBuildNode = {
+  data: ResidualBuildNodeData,
+  kind: void | ResidualBuildKind,
+  type: ResidualBuildNodeType,
+};
+
+// TODO: gradually remove all these, currently it's a random bag of values
+// that should be in args or in other places rather than here.
+export type ResidualBuildNodeData = {
+  appendLastToInvariantBuildNode?: ResidualBuildNode,
+  binding?: Binding | PropertyBinding,
+  boundName?: BabelNodeIdentifier,
+  callTemplate?: () => BabelNodeExpression,
+  concreteComparisons?: Array<Value>,
+  desc?: Descriptor,
+  generator?: Generator,
+  generators?: Array<Generator>,
+  id?: string,
+  lh?: BabelNodeVariableDeclaration,
+  op?: any, // TODO: This is a union of Babel operators, refactor to not use "any" at some point
+  prefix?: boolean,
+  path?: Value,
+  propertyGetter?: SupportedGraphQLGetters,
+  propName?: string,
+  propRef?: ReferenceName | AbstractValue,
+  object?: ObjectValue,
+  quasis?: Array<any>,
+  state?: "MISSING" | "PRESENT" | "DEFINED",
+  thisArg?: BaseValue | Value,
+  template?: PreludeGenerator => ({}) => BabelNodeExpression,
+  typeComparisons?: Set<typeof Value>,
+  typeofString?: string,
+  usesThis?: boolean,
+  value?: Value,
+  violationConditionBuildNode?: ResidualBuildNode,
+};
+
+export type ResidualBuildKind = "DERIVED" | "VOID";
```

Comment:
Without historical knowledge, this sounds weird. How about replacing `kind?: ...` with `hasResult?: boolean`?

Reply:
It can be 1 or 3, not a Boolean

Reply:
It can be 1 or 3, not a Boolean